### PR TITLE
Multi currency implementation for Basilisk on Rococo in KodaDot

### PR DESCRIPTION
**Pull request by vikiival**
# Please test everything again

## What's new
- route /snek
- you should see KSM there
- pay for fees in KSM / BSX
- for everything else use KSM
- faucet soon

## What to test:

- Create collection
- Create NFT
- change your fee payment asset to `KSM` (`/asset`)
- In balanceInputs you should see `KSM` otherwise report as bug
- List NFT
- Make Offer
- MultiFee should always tell you correct currency u use ('You are paying fees in {currency}')
- New prefix is `/snek`

## Things I am aware that does not work correctly

- [ ] New balanceinput has strange validation (always red even if the input is correct)
- [ ] Most probably I broke `/bsx` (we will sunset)
- [ ] Format balance is not the same as Money

## 🐍 changelog

- :snake: snake config
- :zap: added snek
- :bug: max royalty value is 99%
- :spaghetti: pasta from bsx
- :snake: added snek to utils
- :zap: query utils for basilisk
- :zap: Button tells you in which currency u pay for fees
- :label: import Codec as type
- :globe_with_meridians: paying fees in has translation
- :zap: create token has cute link to assets
- :zap: utility package for bsx specific calls
- :alien: asset list for bsx
- :tada: asset page
- :construction: Asset Page
- :globe_with_meridians: asset page
- :bug: :label: AssetItem has id
- :zap: AssetList can fetch user
- :globe_with_meridians: better wording for assets
- :construction: Reworking identity store
- :zap: action to fetch balance
- :bug: correctly fetch balance #3587
- :test_tube: :zap: added data-testid
- :test_tube: added test for BasicBalanceInput
- :alien: asset id for bsx
- :zap: Input for Special tokens
- :zap: CreateToken is using KSM
- :tada: BasicMoney is here
- :zap: ChainMoney as the Money
- :bug: BasicMoney has bad filter
- :bug: ChainMoney should send value
- :zap: show balance of the particular token id
- :bug: Invalid state is handled as 0
- :zap: tokenBalanceInput
- :zap: bsx use multicurrency impl
- :bug: Debounce caused a ton of bugs
- :snake: recognize snek as chain with actions
- :mute: Deleted not fancy log
- :zap: Available actions have dynamic props :)
- :construction_worker: account balance will also handle tokens
- :tada: asset store is here
- :zap: useAppolo can reuse also the apolloClient
- :construction_worker: subscribe for token balance
- :construction_worker: :bug: handle if chain has no token pallet
- :bug: autofetching token balance :white_check_mark:
- :zap: function for gettingTokenBalance
- :construction_worker: WIP OfferBalanceInput
- :zap: asset mixin is here
- :top: export TokenProperty
- :construction_worker: WIP TokenBalanceInput is Using store
- :package: @kodadot1/vuex-options@0.0.1-rc.11
- :broom: remove unused functions

**Thank you for your contribution** to the [KodaDot NFT gallery](https://kodadot.xyz).
👇 \_ Let's make a quick check before the contribution.

### PR type

- [ ] Bugfix
- [x] Feature
- [x] Refactoring

### What's new?

- [ ] PR closes #<issue_number>
- [ ] Requires deployment <rubick/magick_issue>
- [ ] <brief_description_of_what_I've_added>

### Before submitting Pull Request, please make sure:

- [ ] My contribution builds **clean without any errors or warnings**
- [ ] I've merged recent default branch -- **main** and I've no conflicts
- [ ] I've tried to respect high code quality standards
- [ ] I've didn't break any original functionality
- [ ] I've posted a screenshot of demonstrated change in this PR

### Optional

- [ ] I've tested it at </rmrk/collection/26902bc2f7c20c546a-1FVG7>
- [ ] I've tested PR on mobile and everything seems works
- [ ] I found edge cases
- [ ] I've written some unit tests 🧪

### CODE REVIEW

**Comment by yangwao** [comment]
It's moving 👀

**Comment by vikiival** [comment]
> It's moving 👀

@yangwao as a CJO today - Chief Joke Officer

**Comment by vikiival** [comment]
Somehow the I can't submit the offer :/

**Comment by vikiival** [comment]
One Bumble for u 😇 

image

**Comment by yangwao** [comment]
u now

**Comment by preschian** [comment]
I just want to confirm, is this ready for review? because in the title and desc still WIP. how about deepscan, deepsource, and codeclimate issues?

**Comment by yangwao** [comment]
> I just want to confirm, is this ready for review?

I guess yes, just need to specify the testing scenario and what to look for, that's in the section in what's new but not super obvious.
Plus like how to obtain resources on Roc and go on from there :)
Basilisk discord has set faucet for this, probably we should drop one on our Kodadot discord

**Comment by vikiival** [comment]
> I just want to confirm, is this ready for review?

Yes gonna fix deepscan soon.

Also updated desc. thanks for the note

**Review by roiLeo — COMMENTED** [review]
Shouldn't it show "KSM" rather than "BSX"
Capture d’écran 2022-08-27 à 1 29 22 PM

Capture d’écran 2022-08-27 à 1 38 15 PM
Balance 0 KSM? 

BSX sandbox will be off after this PR?

**Review by petersopko — COMMENTED** [review]
good:
- [x] switched fees to KSM
- [x] minted [collection](https://deploy-preview-3784--koda-nuxt.netlify.app/snek/collection/2773267381), paid fees in KSM
- [x] minted and listed [NFT](https://deploy-preview-3784--koda-nuxt.netlify.app/snek/gallery/2773267381-1?page=1), paid fees in KSM
- [x] made generous offer on [Morksi pes](https://deploy-preview-3784--koda-nuxt.netlify.app/snek/gallery/3132385849-2?page=1)
- [x] switched fees to BSX
- [x] minted [NFT](https://deploy-preview-3784--koda-nuxt.netlify.app/snek/gallery/2773267381-2), paid fees in BSX ✅ 

what I'm missing: 
- [ ] assets page isn't accessible other than through link / hyperlink added on relevant places -> perhaps we want to have this in navbar? (possibly tracked somewhere)
- [ ] as @roiLeo mentioned, balance in navbar dropdown always shows BSX
- [ ] seems that there are no balance checks when hitting buy
image
- [ ] transfer only shows bsx
image
- [ ] hover popup on collection in /explore shows bsx
image
- [ ] filtering for price range shows bsx
image

**Comment by petersopko** [comment]
For now, in order to get rococo tokens for testing, I've successfully got mine at [Basilisk faucet channel](https://discord.com/channels/844177080005951489/963084917284155452) by typing `/drip yourBsxAddress`

It's limited to 1KSM+100kBSX/24H right now

I guess our own implementation of this is coming soon? 👀

**Comment by yangwao** [comment]
This needs to be merged so user can copy their address 👀
- https://github.com/kodadot/nft-gallery/pull/3853

**Comment by yangwao** [comment]
waiting for green checks be like 👀

File: components/bsx/Gallery/Item/GalleryItem.vue [hunk 1]
```diff
@@ -83,7 +87,10 @@
                       </div>
                       <div v-if="nftRoyalties" class="royalty">
                         ⊆ {{ $t('royalty') }}
-                        <Money :value="nftRoyalties" inline />
+                        <Money
+                          :value="nftRoyalties"
+                          inline
+                          :prefix="urlPrefix" />
```
Comment (Jarsen136) [inline components/bsx/Gallery/Item/GalleryItem.vue:93]:
```suggestion
       />
```
`prefix` has not been used

**Comment by vikiival** [comment]
https://user-images.githubusercontent.com/22471030/187081275-46ba83e9-8f72-4115-84c8-2c44ffd9bcdc.MOV

**Review by Jarsen136 — COMMENTED** [review]
Just small stuff

**Comment by vikiival** [comment]
@Jarsen136 can you please checkout the branch and try the `pnpm install` ? Looks like my setup is broken :)

**Comment by petersopko** [comment]
> @Jarsen136 can you please checkout the branch and try the `pnpm install` ? Looks like my setup is broken :)

I just got this beauty in console 👀 
image

**Comment by Jarsen136** [comment]
> @Jarsen136 can you please checkout the branch and try the `pnpm install` ? Looks like my setup is broken :)

Some error occurs after I run `pnpm install` at this branch : )
image

**Comment by yangwao** [comment]
`main` builds quite well for me. 
Suspicious is @vikiival has old pnpm and resolved lock file somehow wrongly

**Comment by yangwao** [comment]
Screenshot 2022-08-28 at 18 19 03

Screenshot 2022-08-28 at 18 17 23

**Comment by yangwao** [comment]
Fixed tests are incoming I've heard.

**Comment by vikiival** [comment]
@petersopko do I need to do anything special? 

Screenshot 2022-08-28 at 19 06 55

**Comment by petersopko** [comment]
> @petersopko do I need to do anything special? 
> 
> 
> 
> Screenshot 2022-08-28 at 19 06 55
> 
> 

It's expecting to find text containing "BSX" in that div. Feel free to turn off that test for now, I'll make PR fixing it tomorrow.

File: tests/unit-testing/components/shared/BasicBalanceInput.test.ts [hunk 2]
```diff
@@ -0,0 +1,68 @@
+import { createLocalVue, mount } from '@vue/test-utils'
+import Buefy from 'buefy'
+import { expect, it, vi } from 'vitest'
+
+import BasicBalanceInput from '@/components/shared/form/BasicBalanceInput.vue'
+
+const localVue = createLocalVue()
+localVue.use(Buefy)
+
+let wrapper, input, option
```
Comment (preschian) [inline tests/unit-testing/components/shared/BasicBalanceInput.test.ts:10]: unused `option`?

File: store/assets.ts [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+export const mutations: MutationTree<TokenState> = {
+  SET_ASSET_LIST(state: TokenState, data: any) {
+    state.tokenMap = Object.assign({}, data)
+  },
+}
+
+export const actions: ActionTree<TokenState, TokenState> = {
+  async fetchAssetList(
+    { commit, state }: { commit: Commit; state: TokenState },
```
Comment (preschian) [inline store/assets.ts:32]: unused `state`?

File: store/identity.ts [hunk 4]
```diff
@@ -85,38 +132,63 @@ export const actions = {
       consola.error('[FETCH IDENTITY] Unable to get identity', e)
     }
   },
-  async fetchBalance({ commit, dispatch, rootState }, address: string) {
-    const endpoint = rootState.setting.apiUrl
+  async fetchBalance(
```
[the remark this answers is not in the document]
Comment (preschian) [reply store/identity.ts:135]:
```diff
- async fetchBalance
+ fetchBalance
```

File: components/bsx/specific/MultiPaymentFeeButton.vue [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    try {
+      const api = await this.useApi()
+      this.tokenId = await getAssetIdByAccount(api, this.accountId)
+    } catch (e) {
+      console.log(e)
+    }
+  }
+
+  @Watch('accountId', { immediate: true })
+  async onAccountIdChange(val: string, oldVal: string) {
```
Comment (preschian) [inline components/bsx/specific/MultiPaymentFeeButton.vue:46]: doesn't need async

File: components/bsx/specific/MultiPaymentFeeButton.vue [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+  get url(): string {
+    return `/${this.prefix}/assets`
+  }
+
+  async fetchCurrency() {
+    try {
+      const api = await this.useApi()
+      this.tokenId = await getAssetIdByAccount(api, this.accountId)
+    } catch (e) {
+      console.log(e)
```
Comment (preschian) [inline components/bsx/specific/MultiPaymentFeeButton.vue:41]: not using `$consola`?

File: components/shared/format/BasicMoney.vue [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+export default class BasicMoney extends Vue {
+  @Prop({ default: '0' }) readonly value!: number | string | undefined
+  @Prop({ type: Number, default: 12 }) readonly decimals!: number
+  @Prop({ type: String, default: 'KSM' }) readonly unit!: string
+
+  // MISC
+  @Prop(Boolean) readonly inline!: boolean
+  @Prop(Boolean) readonly hideUnit!: boolean
+
+  // TODO Decimals not properly implemented
```
Comment (preschian) [inline components/shared/format/BasicMoney.vue:28]: link to existing issue or open new issue regarding this one?

File: components/bsx/Asset/AssetList.vue [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+        notificationTypes.success
+      )
+      this.fetchCurrentToken()
+      this.fetchAccountBalance()
+    })
+  }
+
+  @Watch('accountId', { immediate: true })
+  public onAccountIdChange(value: string, oldValue: string) {
+    console.log('onAccountIdChange', value, oldValue)
```
Comment (preschian) [inline components/bsx/Asset/AssetList.vue:99]: is `console.log` safe to remove?

Comment (yangwao) [reply components/bsx/Asset/AssetList.vue:99]: or rather use consola 👀

**Comment by yangwao** [comment]
Screenshot 2022-08-29 at 14 13 50


https://calamar.play.hydration.cloud/rococo%20basilisk/extrinsic/0000715255-000002-0579c



https://calamar.play.hydration.cloud/rococo%20basilisk/extrinsic/0000715279-000002-98663

https://deploy-preview-3784--koda-nuxt.netlify.app/snek/gallery/1559871636-1?message=congrats


image

**Comment by yangwao** [comment]
Okay WFM, let's put it to `main`

**Comment by roiLeo** [comment]
will it blend?

**Pull request merged**